background_scripts/JobQueue.js

```diff
@@ -0,0 +1,118 @@
+class JobDemon {
+  /**
+   * @param {string} url Url d'un job
+   */
+  constructor(url) {
+    /** @type {string} */
+    this.url = url;
+    /** @type {HTMLIFrameElement} */
+    this.iframe = document.createElement("iframe");
+    /** @type {number} */
+    this.lastExecutionTimestamp = Date.now();
+
+    // On charge l'URL dans une iframe
+    // Ici on laisse la main au content script qui va vérifier si un RDV est disponible
+    this.iframe.src = url;
+
+    document.body.appendChild(this.iframe);
+  }
+
+  retry() {
+    this.iframe.contentWindow.postMessage({ type: "retry" }, "*");
+    this.lastExecutionTimestamp = Date.now();
+  }
+
+  kill() {
+    this.iframe.remove();
+    this.iframe = null;
+  }
+}
+
+class JobQueue {
+  /**
+   * @param {number} delayBetweenJobs Delais entre deux jobs, en secondes
+   * @param {number} delayRetryJob Delais entre deux executions d'un même jobs, en secondes
+   * @param {(job: string) => void} onJobStart Callback quand un job débute
+   */
+  constructor(delayBetweenJobs, minDelayRetryJob, onJobStart) {
+    /** @type {number} en millisecondes */
+    this.delayBetweenJobs = delayBetweenJobs * 1000;
+    /** @type {number} en millisecondes */
+    this.delayRetryJob = minDelayRetryJob * 1000;
+    /** @type {(job: string) => void} Callback quand un job débute */
+    this.onJobStart = onJobStart;
+    /** @type {number} */
+    this.intervalRef = null;
+    /** @type {Object<string, JobDemon>} map de deamon traitant les jobs */
+    this.demons = {};
+    /** @type {string[]} tableau d'url de job */
+    this.jobs = [];
+  }
+
+  /**
+   * @param {string} job Url d'un job
+   */
+  add(job) {
+    if (this.jobs.includes(job)) return;
+
+    this.jobs.push(job);
+  }
+
+  /**
+   * @param {string} job Url d'un job
+   */
+  remove(job) {
+    // Supprime un potentiel job existants
+    if (this.jobs.includes(job)) {
+      this.jobs.splice(this.jobs.indexOf(job), 1);
+    }
+  }
+
+  /**
+   * @param {string} job Url d'un job
+   */
+  kill(job) {
+    // Supprimer les deamons existants
+    if (!this.demons.hasOwnProperty(job)) return;
```

Comment:
Petite remarque (encore une fois, ne change pas ici, c'est plus pour l'info personnelle): généralement on évite d'appeler `hasOwnProperty` directement: en effet, rien n'empêche qu'elle ait elle-même été changée sur `demons`...
On va généralement utiliser `Object.prototype.hasOwnProperty.call` mais c'est un peu lourdingue, ET rien n'empêche non plus que ça ait été modifié sur `Object`.

La solution ultime est d'initialiser `this.jobs` avec `this.jobs = Object.create(null)` comme ça son prototype est null et c'est vraiment un objet "pur". Et après on peut juste utiliser `in`: `if (job in this.demons)`.

Tout ça est overkill dans notre contexte (comme l'utilisation de `hasOwnProperty` AMA, mais maintenant que c'est là tu peux le laisser), mais dans des cas où tu dois programmer manière défensive, par exemple pour une librairie, `Object.create(null)` est un bon outil.

Reply:
Bon à savoir, thx